File: Add_two_numbers.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        ListNode *curr1 = l1;
        ListNode *curr2 = l2;
        while (curr1 ->next != NULL && curr2 ->next != NULL){
            curr1 ->val += curr2 ->val;
            curr1 = curr1 ->next;
            curr2 = curr2 ->next;
        }
        if (curr1 ->next == NULL){
            curr1 ->val += curr2 ->val;
            curr1 ->next = curr2 ->next;
        }
        else if (curr2 ->next == NULL){
            curr1 ->val += curr2 ->val;
        }
        curr1 = l1;
        while (curr1 ->next != NULL){
            if (curr1 ->val >9){
                curr1 ->next ->val += curr1 ->val/10;
                curr1 ->val -= curr1 ->val/10*10;
            }
            curr1 = curr1 ->next;
        }
        if (curr1 ->val > 9){
            ListNode *n = new ListNode(curr1 ->val/10);
            curr1 ->val -= curr1 ->val/10*10;
            curr1 ->next = n;
        }
        return l1;
    }
};

File: new_version/add_two_numbers.rb
# =====================
# simplely scan two lists. Problem is the process at the end of list
# =====================

# Definition for singly-linked list.
class ListNode
	attr_accessor :val, :next
	def initialize(val)
		@val = val
		@next = nil
	end
end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
	temp = l1
	is_forward = false
	while temp
		temp.val += l2 ? l2.val : 0
		temp.val += 1 if is_forward
		if temp.val >= 10
			is_forward = true
			temp.val -= 10
		else
			is_forward = false
		end

		l2 = l2.next if l2
		if !temp.next
			temp.next = l2
			l2 = nil
		end
		if !temp.next && is_forward
			temp.next = ListNode.new 1
			is_forward = false
		end
		temp = temp.next
	end
	l1
end

def arr2list(arr)
	temp = nil
	arr.reverse.each do |e|
		current = ListNode.new e
		current.next = temp
		temp = current
	end
	temp
end

l1 = [2, 4, 3]
l2 = [7, 0, 8]
puts add_two_numbers(arr2list(l1), arr2list(l2)).inspect
